system_metrics_collector/src/topic_statistics_collector/received_message_period.hpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TOPIC_STATISTICS_COLLECTOR__RECEIVED_MESSAGE_PERIOD_HPP_
+#define TOPIC_STATISTICS_COLLECTOR__RECEIVED_MESSAGE_PERIOD_HPP_
+
+#include <chrono>
+#include <string>
+
+#include "topic_statistics_collector.hpp"
+
+#include "../../src/system_metrics_collector/collector.hpp"
+
+namespace topic_statistics_collector
+{
+constexpr const char kReceivedMessagePeriodMetricName[] = "received_message_period";
+constexpr const char kUnits[] = "milliseconds";
+constexpr const std::chrono::high_resolution_clock::time_point kDefaultTimePoint{std::chrono::
+  high_resolution_clock::duration::zero()};
+
+template<typename T>
+class ReceivedMessagePeriod : public TopicStatisticsCollector<T>
+{
+public:
+  /**
+   * Constructs a ReceivedMessagePeriod object.
+   *
+   * This also starts the Collector.
+   */
+  ReceivedMessagePeriod()
+  {
+    system_metrics_collector::Collector::Start();
+  }
+  /**
+   * Destructs a ReceivedMessagePeriod object.
+   *
+   * This also stops the Collector.
+   */
+  virtual ~ReceivedMessagePeriod()
+  {
+    system_metrics_collector::Collector::Stop();
+  }
+
+  /**
+   * Handle a message received and measure its received period.
+   * @param received_message
+   */
+  void OnMessageReceived(const T & received_message) override
+  {
+    auto now = GetCurrentTime();
+
+    if (time_last_message_received_ == kDefaultTimePoint) {
+      time_last_message_received_ = now;
+    } else {
+      int period = std::chrono::duration_cast<std::chrono::milliseconds>(
+        now - time_last_message_received_).count();
+      time_last_message_received_ = now;
+      system_metrics_collector::Collector::AcceptData(static_cast<double>(period));
+    }
+  }
+
+protected:
+  /**
+   * Resets the time_last_message_received_ member.
+   * @return
+   */
+  bool SetupStart() override
+  {
+    time_last_message_received_ = kDefaultTimePoint;
+  }
+
+  /**
+   * No-op method
+   * @return
+   */
+  bool SetupStop() override
+  {
+  }
+
+  /**
+   * Returns the current time using high_resolution_clock.
+   * @return the current high_resolution_clock clock time
+   */
+  virtual std::chrono::high_resolution_clock::time_point GetCurrentTime()
```

Comment:
```suggestion
  virtual std::chrono::high_resolution_clock::time_point GetCurrentTime() const
```

Reply:
Add a comment explaining that this is made to allow mocking only, and should never be overridden in production code.

Reply:
Seems like `rclcpp` is embedding an API for clock management. A way to do dependency injection properly here, would be to have the constructor expect a `ros::Clock` object, and if no object is passed, default to a steady clock using the ROS 2 API.

See e.g. rclcpp `test_time.cpp`.

```c++
TEST(TestTime, time_sources) {
  using builtin_interfaces::msg::Time;
  rclcpp::Clock ros_clock(RCL_ROS_TIME);
  Time ros_now = ros_clock.now();
  EXPECT_NE(0, ros_now.sec);
  EXPECT_NE(0u, ros_now.nanosec);

  rclcpp::Clock system_clock(RCL_SYSTEM_TIME);
  Time system_now = system_clock.now();
  EXPECT_NE(0, system_now.sec);
  EXPECT_NE(0u, system_now.nanosec);

  rclcpp::Clock steady_clock(RCL_STEADY_TIME);
  Time steady_now = steady_clock.now();
  EXPECT_NE(0, steady_now.sec);
  EXPECT_NE(0u, steady_now.nanosec);
}
```

The test would then mock the clock class.

Reply:
I'm hesitant to add a direct dependency to ROS2 specific components given this class simply measures a period when a callback is executed. Looking ahead, this class could be part of the library used by the topic statistics implementation, where the lib could be pure c++ and not depend on ROS2.

Reply:
ROS time management has depth to it and we're far better off receiving time than measuring it ourselves, however you want to manage that. For example the `rclcpp::Clock` measures time via the `/clock` topic if the parameter `/use_sim_time` is set on the node, which allows Gazebo or rosbag playback to control the clock.

Reply:
> `rclcpp::Clock` measures time via the `/clock` topic if the parameter `/use_sim_time` is set on the node, which allows Gazebo or rosbag playback to control the clock.

That's good to know. This was also helpful: https://answers.ros.org/question/287946/ros-2-time-handling/